c/CMakeLists.txt

```
@@ -36,9 +36,21 @@ cmake_minimum_required(VERSION 3.19)
 # Project declaration needs to be before the configuration
 project(nvtx3 VERSION 3.1.1)
 
-option(NVTX3_INSTALL "Install NVTX3" ON)
+# Check if standalone or part of another project:
+set(NVTX_C_ROOT "${CMAKE_CURRENT_SOURCE_DIR}")
+string(REGEX REPLACE "/c$" "" NVTX_PROJ_ROOT "${NVTX_C_ROOT}")
+set(NVTX_TESTS_ROOT "${NVTX_PROJ_ROOT}/tests")
+set(NVTX_TOPLEVEL_PROJECT OFF)
+if ("${CMAKE_SOURCE_DIR}" STREQUAL "${NVTX_PROJ_ROOT}" OR
+    "${CMAKE_SOURCE_DIR}" STREQUAL "${NVTX_C_ROOT}" OR
+    "${CMAKE_SOURCE_DIR}" STREQUAL "${NVTX_TESTS_ROOT}")
+  set(NVTX_TOPLEVEL_PROJECT ON)
+endif()
+
+option(NVTX3_TARGETS_NOT_USING_IMPORTED "Turn off to mark NVTX targets as IMPORTED." ${NVTX_TOPLEVEL_PROJECT})
+
+option(NVTX3_INSTALL "Install NVTX3" ${NVTX_TOPLEVEL_PROJECT})
```

I tried setting `NVTX3_INSTALL` without setting `NVTX3_TARGETS_NOT_USING_IMPORTED` and I got an error:
```
CMake Error at build/_deps/nvtx3-src/c/CMakeLists.txt:88 (install):
  install TARGETS given target "nvtx3-c" which does not exist.
```

Should we default `NVTX3_TARGETS_NOT_USING_IMPORTED` to the value of `NVTX3_INSTALL` rather than the value of `NVTX_TOPLEVEL_PROJECT`? 
```suggestion
option(NVTX3_INSTALL "Install NVTX3" ${NVTX_TOPLEVEL_PROJECT})

option(NVTX3_TARGETS_NOT_USING_IMPORTED "Turn off to mark NVTX targets as IMPORTED." ${NVTX3_INSTALL})
```

---

It's tricky to make CMake options dependent on each other like this since they're cached, so changes after the initial configuration won't update dependent options. It's usually a better idea to emit an error if conflicting options are specified.

Beyond that, I don't know enough about the usecases for these options to make the call. Is it expected that users would want the imported targets and also want to install? Or would install rules only be generated when building NVTX as a standalone project?

I note that the current recipe for using NTVX as a user doesn't provide either option, it always builds imported and doesn't provide the ability to use install rules, so maybe the safest path would be to keep the status quo and only define these options when NVTX is a top-level project, and when used inside another project force imported targets + no install.

---

> Is it expected that users would want the imported targets and also want to install?

Generally, no.

IMO, if NVTX is found using `CMakeLists.txt`, it should generally be "built" in the sense of having non-imported targets that can be installed. That's how it would be if we were actually compiling a C/C++ project - CPM finds the sources and builds it, creating non-imported targets. This is the case whether NVTX is the top-level project or not.

If NVTX is found with `find_package()`, then whatever targets the config file creates should be imported. This is already handled automatically with `install(EXPORT)` and friends.

---

> if NVTX is found using CMakeLists.txt, it should generally be "built" in the sense of having non-imported targets that can be installed.

This would be a change to the way the project is currently used. I wouldn't mind making this change if the NVTX3 folks are okay with that. If there's a usecase out there that required the current imported target logic we should make sure we don't break them. @evanramos-nvidia thoughts?

---

I am trying to understand the above. Does this mean that `NVTX3_TARGETS_NOT_USING_IMPORTED` should always be `ON` by default? @KyleFromNVIDIA what you're saying makes sense, I am just trying to make sure I am translating your comment into what we need to change for this PR.

---

Yes, that's what I'm thinking.

---

Okay, let's do this:
```suggestion
option(NVTX3_TARGETS_NOT_USING_IMPORTED "Turn off to mark NVTX targets as IMPORTED." ON)

option(NVTX3_INSTALL "Install NVTX3" ${NVTX_TOPLEVEL_PROJECT})
```